src/main/java/org/embulk/input/jira/JiraInputPlugin.java

```diff
@@ -0,0 +1,233 @@
+        int getTimeoutMillis();
+
+        @Config("retry_limit")
+        @ConfigDefault("5")
+        public int getRetryLimit();
+
+        @Config("jql")
+        public String getJQL();
+
+        @Config("columns")
+        public SchemaConfig getColumns();
+
+        // For future support of other authentication methods
+        @Config("auth_method")
+        @ConfigDefault("\"basic\"")
+        public AuthenticateMethod getAuthMethod();
+    }
+
+    @Override
+    public ConfigDiff transaction(ConfigSource config,
+            InputPlugin.Control control)
+    {
+        PluginTask task = config.loadConfig(PluginTask.class);
+
+        Schema schema = task.getColumns().toSchema();
+        int taskCount = 1;
+
+        return resume(task.dump(), schema, taskCount, control);
+    }
+
+    @Override
+    public ConfigDiff resume(TaskSource taskSource,
+            Schema schema, int taskCount,
+            InputPlugin.Control control)
+    {
+        control.run(taskSource, schema, taskCount);
+        return Exec.newConfigDiff();
+    }
+
+    @Override
+    public void cleanup(TaskSource taskSource,
+            Schema schema, int taskCount,
+            List<TaskReport> successTaskReports)
+    {
+    }
+
+    @Override
+    public TaskReport run(TaskSource taskSource,
+            Schema schema, int taskIndex,
+            PageOutput output)
+    {
+        PluginTask task = taskSource.loadTask(PluginTask.class);
+        JiraUtil.validateTaskConfig(task);
+        JiraClient jiraClient = getJiraClient();
+        jiraClient.checkUserCredentials(task);
+        try (final PageBuilder pageBuilder = getPageBuilder(schema, output)) {
+            if (isPreview()) {
+                List<Issue> issues = jiraClient.searchIssues(task, 0, PREVIEW_RECORDS_COUNT);
+                for (Issue issue : issues) {
+                    issue.toRecord();
+                    JiraUtil.addRecord(issue, schema, task, pageBuilder);
+                }
+            }
+            else {
+                int currentPage = 0;
+                int totalCount = jiraClient.getTotalCount(task);
+                int totalPage = JiraUtil.calculateTotalPage(totalCount, MAX_RESULTS);
+                LOGGER.info(String.format("Total pages (%d)", totalPage));
+                while (currentPage < totalPage) {
+                    LOGGER.info(String.format("Fetching page %d/%d", (currentPage + 1), totalPage));
+                    List<Issue> issues = jiraClient.searchIssues(task, (currentPage * MAX_RESULTS), MAX_RESULTS);
+                    for (Issue issue : issues) {
+                        issue.toRecord();
+                        JiraUtil.addRecord(issue, schema, task, pageBuilder);
+                    }
+                    currentPage++;
+                }
+            }
+            pageBuilder.finish();
+        }
+        return Exec.newTaskReport();
+    }
+
+    @Override
+    public ConfigDiff guess(ConfigSource config)
```

Comment:
Looks like we inherit a very small portion of "embulk-parser-jsonpath" (which actually just delegate to embulk's built-in's `SchemaGuess`). Writing our own ruby function to leverage `SchemaGuess` seems like a leaner way comparing to this. Our own version for the ruby function will be even shorter, we know exactly where is the JSON root, whether it's an array or object and how the guess result diff should be shaped.
This is just a suggestion, but please consider, I think could save later maintainers some time for they need to figure out why an input plugin directly requires a parser plugin.

Reply:
Sincerely I would enjoy the code from the other plugin.
Write it is not a problem but we could potential duplicate code from both https://github.com/treasure-data/embulk-input-jira/blob/master/lib/embulk/input/jira_input_plugin_utils.rb#L30 and https://github.com/embulk/embulk/blob/57b42c31d1d539177e1e818f294550cde5b69e1f/lib/embulk/guess/schema_guess.rb#L16-L24%20def%20from_hash_records(array_of_hash)%20array_of_hash%20=%20Array(array_of_hash)%20if%20array_of_hash.empty?%20raise%20%22SchemaGuess%20Can't%20guess%20schema%20from%20no%20records%22%20end%20column_names%20=%20array_of_hash.map(&:keys).inject([])%20{|r,%20a|%20r%20+%20a%20}.uniq.sort%20samples%20=%20array_of_hash.to_a.map%20{|hash|%20column_names.map%20{|name|%20hash[name]%20}%20}%20from_array_records(column_names,%20samples)%20end%20end%20end%20end%20module%20Embulk%20module%20Input%20module%20JiraInputPluginUtils%20#%20Guess::SchemaGuess.from_hash_records%20returns%20Columns%20#%20containing%20'index'%20key,%20but%20it%20is%20needless.%20def%20self.guess_columns(records)%20schema%20=%20Guess::SchemaGuess.from_hash_records(records)%20schema.map%20do%20|c|%20column%20=%20{name:%20c.name,%20type:%20c.type}%20column[:format]%20=%20c.format%20if%20c.format%20column%20end%20end%20def%20self.cast(value,%20type)%20return%20value%20if%20value.nil?%20case%20type.to_sym%20when%20:long%20Integer(value)%20when%20:double%20Float(value)%20when%20:timestamp%20Time.parse(value)%20when%20:boolean%20!!value%20else%20value.to_s%20end%20end%20end%20end%20end
So I decide to use the existing one.
As https://github.com/embulk/embulk/blob/v0.9.12/embulk-core/src/main/java/org/embulk/spi/FileInputRunner.java#L73-L86 also do the same

Reply:
Your call, this is just a suggestion. I don't think we're on the same page, but let's take the discussion offline.